rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+ │                     ┌───▼─────┐        ┌──▼──────┐                      │
+ │                     │Handshake│        │Handshake├───────────────┐      │
+ │ ┌───────────────────┤ Success │        │ Failed  │               │      │
+ │ │                   └───┬─────┘        └┬────────┘               │      │
+ │ │                       │               │                        │      │
+ │ │Negotiated             │No ALPN        │remote error: tls:      │other │
+ │ │teleport-reversetunnel │Negotiated     │no application protocol │errors│
+ │ │                       │               │                        │      │
+ │ │                   ┌───▼────┐          │                  ┌─────▼──┐   │
+ │ │                   │Upgrade ◄──────────┘                  │  Not   │   │
+ └─┼───────────────────►Required│                             │Required◄───┘
+   │                   └────────┘                             └────▲───┘
+   │                                                               │
+   └───────────────────────────────────────────────────────────────┘
+```
+Connection upgrade for TLS Routing should be required when the client and the server fail to neogiate the ALPN protocol,
+hinting there is a load balancer in the middle that terminates TLS.
+
+However, this test may not be bullet proof. In particular, the test explicitly looks for `tls: no application protocol`
+from the remote when handshakes fails, but it is possible that a load balancer implementation decides to use a different
+text.
+
+In such cases, setting the mode to `on` can enforce the clients to always do the connection upgrade. Also, setting the
+mode `on` or `off` improves performance by skipping the auto test. An environment variable (e.g.
+`TELEPORT_TLS_ROUTING_UPGRADE_MODE=off`) can be set to overwrite the mode per client, just in case.
+
+### Supporting all TLS Routing protocols
+
+In general, in ALL existing places we are dialing the TLS Routing connection, common logic will be added to the
+`ContextDialer` to automatically upgrade the connection when necessary.
+
+However, not all connections are initiated by `tsh`, `teleport`, or a client usingthe API lib. In such cases, a [local
+Teleport
+proxy](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md#local-teleport-proxy)
+is required to dial the TLS Routing connection with the upgrade.
+
+Local proxy has already been implemented for Database Access and Application Access. Here are some details how it can be
+implemented for Kubernetes Access.
+
+#### Local proxy for Kubernetes Access
+
+`tsh` will serve a local HTTPS proxy server with a localhost self-signed CA. An ephemeral (per `tsh` command) KUBECONFIG
+will be provided to Kubernetes applications to connect to the local server. This KUBECONFIG points to the local proxy
+using the cluster `server` field, and the config will include static user credentials generated with the local
```

Comment:
`kubeconfig` has a field to specify the `proxy` URL using `proxy-url`. 

If we know that the cluster is using the ALB routing mode, we can force that field. It will help `kubectl`  to print a better error message if the proxy is not running locally instead of trying to dial a cluster that he does not know the expected protocol. For other use cases such as creating an MFA session without being prompted to introduce the MFD, we can use the `HTTPS_PROXY` env var

This change should be applied to teleport clusters and not to other direct clusters the user might have configured in  his kubeconfig

Reply:
Any advantage to using `proxy-url` to point to the local proxy over using `server` to point to the local proxy directly, for teleport clusters?

Reply:
Not really. 
When I initially thought about the issue I imagined it using the default `.kube/config` file and on that, it would be better to use the `proxy-url`. 

But I got your idea of copying the default kubeconfig, populating it with `tsh proxy kube <cluster1> <cluster2>` and storing it under `/tmp/kubeconfig-adhahdhashd.yaml`  with the server being the local server and the SNI identifying the cluster, Once `tsh proxy kube` starts you sign the cluster certificates and negotiate the MFA session.

In that scenario, we don't need the `proxy-url`.